### PR TITLE
Docker secrets support

### PR DESCRIPTION
**Pull request by jason-fox**
Moves the defaults for username and password out of the `Dockerfile` and into the `entrypoint.sh` where they can be read in directly (as before) or via a file uploaded via [Docker Secrets](https://docs.docker.com/engine/swarm/secrets/)

In summary, to improve security, it should be possible to set all `ENV` variables for Username and password without passing plaintext.  It therefore won't matter if someone does a `docker inspect` the sensitive information won't be available - merely a path to a secrets file.

This basically a copy-paste from the official MySQL docker files and uses the de-facto standard naming convention.

## Docker Secrets

As an alternative to passing sensitive information via environment variables, `_FILE` may be appended to the listed environment variables, causing the initialization script to load the values for those variables from files present in the container. In particular, this can be used to load passwords from Docker secrets stored in `/run/secrets/<secret_name>` files.

* `CYGNUS_MYSQL_USER_FILE`
* `CYGNUS_MYSQL_PASS_FILE`
* `CYGNUS_MONGO_USER_FILE`
* `CYGNUS_MONGO_PASS_FILE`
* `CYGNUS_HDFS_USER_FILE`
* `CYGNUS_HDFS_TOKEN_FILE`
* `CYGNUS_POSTGRESQL_USER_FILE`
* `CYGNUS_POSTGRESQL_PASS_FILE`
* `CYGNUS_CARTO_USER_FILE`
* `CYGNUS_CARTO_KEY_FILE`

For example

```
docker run --name some-cygnus -e CYGNUS_MYSQL_PASS_FILE=/run/secrets/mysql-root -d fiware/cygnus-ngsi:tag
```

### CODE REVIEW

**Comment by AlvaroVega** [comment]
I'm not agree on remove old env vars, backward compatibility should be provided.

**Comment by jason-fox** [comment]
Backwards compatibility **is** provided - it **doesn't** remove the old vars. It just sets up the defaults later in `entrypoint` ( using `$2` in the function) rather than hardcoding the defaults in `Dockerfile` 

The default values are now set to `''` **later** in `entrypoint` since you can't set both  `CYGNUS_MYSQL_PASS` and `CYGNUS_MYSQL_PASS_FILE` simultaneously. There a check to avoid duplicates. If the Dockerfile would set the default you could never set the `_FILE` instead.

Look at MySQL for another example of adding secrets support - they haven't removed the old parameters, just added the option for the new ones.

**Comment by AlvaroVega** [comment]
I was looking at: https://github.com/telefonicaid/fiware-cygnus/pull/1503/files#diff-d0fccc3e1aa657c810aaa0ea13186de7R40

**Comment by jason-fox** [comment]
I left commented out defaults in `Dockerfile` for informational purposes, since as a developer when I read the `Dockerfile` I'd like to see which ENV variables are supported - for example  `CYGNUS_MYSQL_USER` remains as supported variable with a default value of `''`

However in entrypoint, the line: `file_env 'CYGNUS_MYSQL_USER' ''` is now doing the same job as the old `ENV CYGNUS_MYSQL_USER ""` in the Dockerfile - it is:

* Checking to see if `CYGNUS_MYSQL_USER` is already set and failing if `CYGNUS_MYSQL_USER_FILE` is set as well
* If only `CYGNUS_MYSQL_USER_FILE` is set, the value is read from the file and an env for `CYGNUS_MYSQL_USER` is set.
* If neither is `CYGNUS_MYSQL_USER` or `CYGNUS_MYSQL_USER_FILE` is set `CYGNUS_MYSQL_USER`  is set to the default value - `''`

This covers all the possibilities without changing existing functionality

**Comment by jason-fox** [comment]
See the following security issue: moby/moby/issues/13490 for reasoning behind adding this

**Comment by AlvaroVega** [comment]
Thanks @jason-fox 
LGTM

File: docker/cygnus-ngsi/Dockerfile [hunk 1]
```diff
@@ -37,8 +37,8 @@ ENV CYGNUS_MULTIAGENT false
 # MySQL options
 ENV CYGNUS_MYSQL_HOST ""
 ENV CYGNUS_MYSQL_PORT ""
-ENV CYGNUS_MYSQL_USER ""
-ENV CYGNUS_MYSQL_PASS ""
+# ENV CYGNUS_MYSQL_USER ""
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply docker/cygnus-ngsi/Dockerfile]: Commented intentionally - Personally when I read the `Dockerfile` it is useful to be able to see the all the **supported** `ENV` variables listed at the top.

Until now `CYGNUS_MYSQL_USER` `CYGNUS_MYSQL_PASS` etc have been in that list and indeed you can still use them - hence they've been left in the file.

However it **may** make sense to group them together and  highlight that using these in a Production environment is insecure - I'm not sure what best practise would be

I'm open to suggestions:

* Leave - User can read them in the Dockerfile.
* Delete - User must read about them in the Documentation
* Group together and add Insecure Warning - User is informed with both

Which would you prefer?

[the remark this answers is not in the document]
Comment (fgalan) [reply docker/cygnus-ngsi/Dockerfile]: As far as I understand, the idea of having the `ENV CYGNUS_<envar> ""` lines in the Dockerfile is to keep the reference of the ones used by the component so local builds can edit the file in order to provide direclty the configuration.

With this in mind, my suggestion is to keep the variable list as it is know but adding a warning about the ones which are insecure. For instance:

```
# MongoDB / STH options
ENV CYGNUS_MONGO_HOSTS ""
ENV CYGNUS_MONGO_USER ""   # Warning: using this env var in clear text is insecure
ENV CYGNUS_MONGO_PASS ""   # Warning: using this env var in clear text is insecure
ENV CYGNUS_MONGO_ENABLE_ENCODING ""
ENV CYGNUS_MONGO_ENABLE_GROUPING ""
ENV CYGNUS_MONGO_ENABLE_NAME_MAPPINGS ""
```
or a more appropiate wording for the warning message.

I thinki is similar to the 3rd option you mention, but without "breaking" the grouping by sink.

[the remark this answers is not in the document]
Comment (jason-fox) [reply docker/cygnus-ngsi/Dockerfile]: Fixed b2f8414f - I annotated both the `Dockerfile` and the readme and kept the ordering as is.

**Pull request merged**